packages/mdc-menu-surface/foundation.ts

```diff
@@ -0,0 +1,500 @@
+    return numbers;
+  }
+
+  static get Corner() {
+    return Corner;
+  }
+
+  /**
+   * {@see MDCMenuSurfaceAdapter} for typing information on parameters and return types.
+   */
+  static get defaultAdapter(): MDCMenuSurfaceAdapter {
+    return {
+      addClass: () => undefined,
+      removeClass: () => undefined,
+      hasClass: () => false,
+      hasAnchor: () => false,
+
+      isElementInContainer: () => false,
+      isFocused: () => false,
+      isFirstElementFocused: () => false,
+      isLastElementFocused: () => false,
+      isRtl: () => false,
+
+      getInnerDimensions: () => ({height: 0, width: 0}),
+      getAnchorDimensions: () => null,
+      getWindowDimensions: () => ({height: 0, width: 0}),
+      getBodyDimensions: () => ({height: 0, width: 0}),
+      getWindowScroll: () => ({x: 0, y: 0}),
+      setPosition: () => undefined,
+      setMaxHeight: () => undefined,
+      setTransformOrigin: () => undefined,
+
+      saveFocus: () => undefined,
+      restoreFocus: () => undefined,
+      focusFirstElement: () => undefined,
+      focusLastElement: () => undefined,
+
+      notifyClose: () => undefined,
+      notifyOpen: () => undefined,
+    };
+  }
+
+  private isOpen_: boolean;
+  private openAnimationEndTimerId_: number;
+  private closeAnimationEndTimerId_: number;
+  private animationRequestId_: number;
+  private dimensions_!: MenuDimensions; // initialized in open()
+  private anchorCorner_: Corner;
+  private anchorMargin_: MenuDistance;
+  private measurements_!: AutoLayoutMeasurements; // initialized in open()
+  private isQuickOpen_: boolean;
+  private isHoistedElement_: boolean;
+  private isFixedPosition_: boolean;
+  private position_: MenuPoint;
+
+  constructor(adapter: MDCMenuSurfaceAdapter) {
+    super(Object.assign(MDCMenuSurfaceFoundation.defaultAdapter, adapter));
+
+    this.isOpen_ = false;
+    this.openAnimationEndTimerId_ = 0;
+    this.closeAnimationEndTimerId_ = 0;
+    this.animationRequestId_ = 0;
+    this.anchorCorner_ = Corner.TOP_START;
+    this.anchorMargin_ = {top: 0, right: 0, bottom: 0, left: 0};
+    this.isQuickOpen_ = false;
+    this.isHoistedElement_ = false;
+    this.isFixedPosition_ = false;
+    this.position_ = {x: 0, y: 0};
+  }
+
+  init() {
+    const {ROOT, OPEN} = MDCMenuSurfaceFoundation.cssClasses;
+
+    if (!this.adapter_.hasClass(ROOT)) {
+      throw new Error(`${ROOT} class required in root element.`);
+    }
+
+    if (this.adapter_.hasClass(OPEN)) {
+      this.isOpen_ = true;
+    }
+  }
+
+  destroy() {
+    clearTimeout(this.openAnimationEndTimerId_);
+    clearTimeout(this.closeAnimationEndTimerId_);
+    // Cancel any currently running animations.
+    cancelAnimationFrame(this.animationRequestId_);
+  }
+
+  /**
+   * @param corner Default anchor corner alignment of top-left menu surface corner.
+   */
+  setAnchorCorner(corner: Corner) {
+    this.anchorCorner_ = corner;
+  }
+
+  /**
+   * @param margin Set of margin values from anchor.
```

Comment:
remove closure annotation. Can keep the rest of the comment though.

Reply:
This is actually [TypeDoc](https://typedoc.org/guides/doccomments/) syntax 😀